File: Go/1657. Determine if Two Strings Are Close.go
package main

import (
	"reflect"
	"sort"
)

//goland:noinspection GoUnusedFunction
func closeStrings(word1 string, word2 string) bool {
	if len(word1) != len(word2) {
		return false
	}

	word1Map := make(map[rune]int)
	word2Map := make(map[rune]int)

	for _, char := range word1 {
		word1Map[char]++
	}

	for _, char := range word2 {
		word2Map[char]++
	}

	if len(word1Map) != len(word2Map) {
		return false
	}

	word1Counts := make([]int, 0, len(word1Map))
	word2Counts := make([]int, 0, len(word2Map))

	for charCount := range word1Map {
		word1Counts = append(word1Counts, word1Map[charCount])
		if _, exists := word2Map[charCount]; !exists {
			return false
		}
		word2Counts = append(word2Counts, word2Map[charCount])
		delete(word2Map, charCount)
	}

	sort.Ints(word1Counts)
	sort.Ints(word2Counts)

	return reflect.DeepEqual(word1Counts, word2Counts)
}

File: JavaScript/1657. Determine if Two Strings Are Close.js
// noinspection JSUnusedGlobalSymbols

/**
 * @param {string} word1
 * @param {string} word2
 * @return {boolean}
 */
const closeStrings = (word1, word2) => {
  if (word1.length !== word2.length) {
    return false
  }

  const word1Map = new Map()
  const word2Map = new Map()

  for (const char of word1) {
    word1Map.set(char, (word1Map.get(char) || 0) + 1)
  }

  for (const char of word2) {
    if (!word1Map.has(char)) {
      return false
    }
    word2Map.set(char, (word2Map.get(char) || 0) + 1)
  }

  const word1Counts = [...word1Map.values()].sort((a, b) => a - b)
  const word2Counts = [...word2Map.values()].sort((a, b) => a - b)

  for (let i = 0; i < word1Counts.length; i++) {
    if (word1Counts[i] !== word2Counts[i]) {
      return false
    }
  }

  return true
}
